crates/biome_js_analyze/src/services/typed.rs

```diff
@@ -120,9 +118,7 @@ impl Visitor for ScopedResolverBuilderVisitor {
 
         match event {
             WalkEvent::Enter(node) => {
-                if let Some(expr) =
-                    JsExpressionStatement::cast_ref(node).and_then(|node| node.expression().ok())
```

Comment:
@arendjr This prevented const declarations and reference expressions to be resolved. Is this condition correct? Changing this to `AnyJsExpression` fixed the issue, but it can occur performance regression.

Reply:
I had indeed avoided that for performance reasons, and this was all I needed for it to work with `noFloatingPromises`... but indeed that doesn't scale across the board :(

I'd be happy to relax this to support `AnyJsExpression`, but I'm afraid we need to resolve the performance issues first.

Reply:
For now, registering types for the discriminant and the case expressions is sufficient for the rule: bdfc07bda2d6a1f3b854a945de7be1906af9501c